### ISSUE TITLE
It took me a long time to load the big map with esnimi.

### ISSUE THREAD

**Issue by liufuda** [esmini/esmini#305]
It took me a long time to load the big map with esnimi.

I found that the following two places took many time to parse the map.

Spend a lot of time in setlaneosipoints() and setroadmarkosipoints() functions to find sampling points.
I think we can confirm the sampling point through reference line,
For example, if geometry is line, we can use the first point and the last point {s_, s_ + length_};
If geometry is arc, samples can be taken every 1 degree
const double s_step = 0.01 / std::abs(this->curvature_); // sample at approx. every 1° std::set<double> s_vals; for (double s = s_; s < (s_ + length_); s += s_step) s_vals.insert(s); s_vals.insert(s_ + length_);
2.GetRoadById.
This function is called in many places. Whether to consider adding a container std:: unordered_ Map<int, road * > to improve the efficiency of this function
Thanks!

**Comment by scpa1055** [comment]
i also observed something similar with the current version 2.25.1.
With this version i am not able to launch esmini anymore.
Although with the older version 2.23.1 everything worked fine.
My xodr is about 32mb!
As for now, during the step "No scenegraph 3D model loaded. Generating a simplistic one..." the process gets timeout and killed due to huge memory usage. (32GB) !
i dont have a osgb file, thats why esmini tries to generated a simplistic one

**Comment by eknabe** [comment]
Would it be possible to share the xodr, or any xodr causing the same error?

ons 27 juli 2022 kl. 15:51 skrev scpa1055 ***@***.***>:

**Comment by scpa1055** [comment]
Sorry for deleting my comment, wasnt ment to do so.
Sorry cant share the xodr because its proprietary.
In my case the breaking change appears in 2.24.0 causing memory leak. Must have something to do with signals and the osgb creation at least i guess...but need to investigate that further.
Maybe i can share a snippet.

**Comment by qqkkwan** [comment]
i met the same problem, my xodr file is about 8MB . Unfortunately it can't be shown by esmini version 2.26，but  [odrviewer](https://odrviewer.io/)  can show, could  you give me an address that i can share to you ,thanks.

**Comment by eknabevcc** [comment]
Yes, it would be very valuable to debug with your xodr. 
Please share (maybe you can zip it). You find my email at bottom of this page: 
https://sites.google.com/view/simulationscenarios

**Comment by eknabevcc** [comment]
> Sorry for deleting my comment, wasnt ment to do so. Sorry cant share the xodr because its proprietary. In my case the breaking change appears in 2.24.0 causing memory leak (ate my 32GB Memory). Must have something to do with signals and the osgb creation at least i guess...but need to investigate that further. Maybe i can share a snippet.

If the issue still is valid, would it be possible to share a snippet? It is possible to share by mail if desired.

**Comment by sterres** [comment]
Hi, 

I also see memory leak with big maps and esminilib + OSI from Python (eats up 32GB RAM very fast).
![image](https://user-images.githubusercontent.com/30345042/219340548-f043e921-20e8-43a2-a1f8-9904b0bfaede.png)


I suggest to use this big OpenDRIVE map for testing as it is available open-source
https://github.com/BMEAutomatedDrive/ZalaZONE-automotive-proving-ground-virtual-simulation-models

BR
Simon

**Comment by eknabevcc** [comment]
When I run [this scenario](https://github.com/esmini/esmini/files/10755460/dummy.xosc.txt) based on [ZalaZONE_full.xodr](https://github.com/BMEAutomatedDrive/ZalaZONE-automotive-proving-ground-virtual-simulation-models/blob/master/OpenDrive/ZalaZONE_full.xodr) in esmini (v2.29.2) the OSI file becomes ~1.3 MB. And  esmini process consuming ~204 MB. 

This is the exact launch command:
`./bin/esmini --window 60 60 800 400 --osc dummy.xosc --osi_file --fixed_timestep 0.05`

It seems I can't reproduce the issue you encounter. Can you try this very command with the same assets (scenario and OpenDRIVE)?

Do you get very different results? In that case let's dig deeper: What system do you run on (OS + version)?

**Comment by sterres** [comment]
You are right - with your exact setup & esmini.exe for me it also works fine.

I can reproduce with the OSI Python example (using esminiLib.dll) and the ZalaZone Map.
https://github.com/esmini/esmini/tree/master/EnvironmentSimulator/code-examples/hello_world/osi_groundtruth.py

I had to change the time step size from 0.05s to 0.01s to have the effect visible (in line 73).

Maybe it's not even an esmini bug but just related to the Python memory handling?

**Comment by eknabevcc** [comment]
Can't reproduce, unfortunately. I get ~215MB RAM usage it and loads fairly quick.

Let's try find out what is the difference in our setup:

- Did you use that very same OpenDRIVE file?
- esmini v2.292? (I assume yes, you already mentioned it)
- What Python version? (I use 3.9.13)
- Windows I assume? (I run on Windows 10)
- Protobuf version? (check with `pip show protobuf`, I have 3.20.1)
- Make sure you're using the latest osi3 python files (from [scripts/osi3](https://github.com/esmini/esmini/tree/master/scripts/osi3))

What more can be relevant to check?

**Comment by sterres** [comment]
Good points....I found the difference to be protobuf. I automatically got version 4.21.12 installed which uses a fast "upd" backend (and also seems to have the memory leak).
When I install protobuf 3.20.1, I see a memory consumption between 250-300MB but it is very much slower with the "python" backend (around factor 10000 slower).

Maybe you can try if you can reproduce the memory leak with the protobuf 4.x? 
It seems not to be an esmini bug then :)
